### DISCUSSION TITLE
How should we adapt our program to resolve new and continuing challenges with demos?

### DISCUSSION THREAD

**Discussion by karenword** [carpentries/instructor-training#1610] [checkout-feedback]
A few of the challenges cited recently are when trainees:
- manage to avoid live coding even when using a recommended episode
- over-populate a prepared notebook
- invite themselves to observe a demo without participating

Please feel free to add to this list of challenges, in addition to proposing solutions.

**Comment by zpainter** [comment]
Something I find, not uncommon with how you could game it, is that many trainees do a good thing by setting up the context of a lesson. They ask certain questions in the beginning, and make sure things are well. I think this is a good teaching practice, but if you are a bit long winded, you can easily burn most of the time up, which might work in your favor if nervous about other things.

Other than making the demo a bit longer, or trying to provide more guidance about how to do the intro to a lesson, I am not sure how to get around this particular one.

**Reply by jevbelikov** [reply]
I often overrun because of extending demo time, so I tend to tell the participant the slot will be an hour and a half.
It also feel that giving/receiving verbal feedback is nicer than writing it only in the doc, but this also takes more time.

**Comment by sstevens2** [comment]
Do they need to do live coding as part of the demo? I think of it as a "teaching demo" not "live coding demo" so even if they don't get to doing live coding I can still review their teaching skill during those 5 minutes.

On the other side, trainees are often nervous about doing live coding and I think having an opportunity to practice is helpful.  I usually let them start wherever they want in the episode.  I've had quite a few people ask me where they should start with the new change.  It happened before with the old system but we had a curated list so starting at the top was usually good then.  Now, I tell them they can start anywhere as long as they feel comfortable starting in that location but recommend starting at the top as it has the least likelihood of having dependencies (which they should prepare for if they start further down).  Maybe we could have some more info text about how to choose a starting place when they pick their episode? If they want to practice more live coding, they may want to start further down in the episode, for example.

**Comment by sstevens2** [comment]
The prepared notebook issue came up for me and I mentioned it a little in the slack thread but will try to summarize here.

It can be a good opportunity to talk about the pros and cons of using a prepared notebook.  Given our usual audience (see below for more detail) for most workshops, seems like it is usually better to start with the tool you are working with nothing pre-populated.  I struggled with if I should ask someone to repeat because otherwise the teaching demo I observed was very good and they were thoughtful about our discussion about pros and cons.

<details>
  <summary>Notebook Pros/Cons aside</summary>
For most of the Library/Data/Software Carpentry curricula, they are novices starting out and so I think it is typically better to start with a blank document for those folks and focus their attention on something simple as you build up to more complexity.  This seems to help with cognitive load and reading the notebook can be distracting.

For intermediate workshops, where you have some coding skill coming in but are building up a particular secondary skill, sometimes giving them a bit of the code in advance can be really useful, to focus them on the new thing they are learning, not the foundation they already know but need to use in learning the new skill.  Also sometimes it requires a lot of code and then you spend a lot of time fixing typos when those are things the learners are already familiar with. For example, we've taught the ML intro workshop in the incubator and some of that has larger sections of code that learners are already familiar with to build up to a new concept that they are learning.

Also think it is important to note if the instructor is using a pre-filled notebook, they need to give it to the learners as well.  This also means spending time talking the learners through downloading it, opening it, what the contents of the notebook are, and how to read/interact with it.
</details>

**Reply by laufers** [reply]
For beginners, I am not a fan of prepared notebooks. I do see a usefulness of a prepared  notebook for an intermediate and expert workshops. I use one for an intermediate matplotlib workshop I teach. It is easier to have the lines of code that generate the data already available and a commented faded section for the actual matplotlib code used for the live coding portion of the workshop. Usually this is is just a cell with comments for where the learner will be coding along with me to fill. If live coding examples are needed later, I will go ahead and have it already in a cell for the learner to then modify as part of the next topic.

**Comment by sstevens2** [comment]
Starting a threaded discussion about over preparing too.  Not sure if this is more evident or more possible with the new system because trainees only prepare 5 minutes, but it seems like there have been a few instances of trainees highly preparing the 5 minutes and then not feeling comfortable teaching outside of it.  This is much more difficult to do when you are teaching a longer session, which may set up unrealistic expectations for them when they go to prepare for a real workshop.

In my demos, I've observed..
-  One person who mentioned they had prepared it exactly but then didn't follow it exactly so they were disappointed in themselves.  They otherwise did a great job!  And in my experience, you are unlikely to teach it exactly the same each time.  As an instructor, I try to be flexible to the learners around me and what I or other instructors have built up so far in the lesson.  It was a good discussion about how it may come out differently than you prepared but that is okay as long as the learners are supported and learn from it.
- Another person who went slightly faster than they expected and then had ~10 seconds of not being sure what to do at the end because they'd only prepared those 5 minutes.  They otherwise did very well and there is slightly more to this story but it generally illustrates the concept of over-preparing only 5 minutes.

Other observations:
- I have noticed a lot more of learners starting their own timers for 5 minutes, maybe because they practiced it this way?
- This might also be related to the notebook topic above a bit as it was probably harder to create a pre-populated notebook when you didn't know which episode you might be teaching from.

**Reply by zpainter** [reply]
I usually tell trainees that I do not do 5 exact minutes. Instead I might go slightly longer, and I frame it as, I do not want to cut you off in the middle of a thought, so keep going! That seems to take the time anxiety down a bit, and gives me the time to see how natural someone is in a session.
